### ISSUE TITLE
Debugging is not working for web with Chrome 100.x update

### ISSUE THREAD

**Issue by pbeccegato** [flutter/flutter#101224]
I’m using Flutter from 2 years on my mac to dev web app without any troubles.
Latest stable 2.10.4
From today, when I launch my app in debug mode, when the code reaches a breakpoint, in VSCode appears the pause symbol, but I can’t see any variables or stack. I’ve tryed in many ways to solve the issue without success. I also tryed with a new sample app; same trouble. Also using Android Studio; same trouble. If I debug on iPhone emulator, debug works fine and on a breakpoint; I can see VSCode jump to the breakpoint line and the variables are inspectable.

When I stop the debug, in console debug appears thie error:
Failed to save Chrome preferences: FileSystemException: Cannot open file, path = '[/var/folders/ps/ryyd46zx367gn2zdtn1_yk4c0000gn/T/flutter_tools.0r3GeL/flutter_tools_chrome_device.obsal2/Default/.com.google.Chrome.hDJYmg]()' (OS Error: No such file or directory, errno = 2)

Any help will be appreciated.
Thanks in advance.
Peter

**Comment by pbeccegato** [comment]
After spent 9 hours to make test I have discover that the problem is the new release 100.0.4896.64 of Chrome. Using previous version 99.0.4844.151, everthing works fine

**Comment by tobefine** [comment]
the  problem  trouble me three days ,this is the only way that i found. thanks

**Comment by maheshmnj** [comment]
Hi @pbeccegatom, Thanks for filing the issue. I am able to reproduce the issue on stable and the master channel. Hitting a breakpoint doesn't much provide any info verified in vscode with the default counter app.


https://user-images.githubusercontent.com/31410839/161542653-bae0628f-b310-4666-bb95-3f469b13016f.mov



<details>
<summary>flutter doctor -v (mac)</summary>

```
[✓] Flutter (Channel stable, 2.10.4, on macOS 12.3 21E230 darwin-arm, locale en-GB)
    • Flutter version 2.10.4 at /Users/mahesh/Documents/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision c860cba910 (3 days ago), 2022-03-25 00:23:12 -0500
    • Engine revision 57d3bac3dd
    • Dart version 2.16.2
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.10.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.65.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.37.20220301

[✓] Connected device (4 available)
    • Redmi K20 Pro (mobile) • 192.168.1.2:55555                    • android-arm64  • Android 11 (API 30)
    • iPhone 12 Pro (mobile) • 535317A4-921C-488F-B234-C2CC6D4AE5A1 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 12.3 21E230 darwin-arm
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 99.0.4844.83

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

```
[✓] Flutter (Channel master, 2.13.0-0.0.pre.357, on macOS 12.3 21E230 darwin-arm, locale en-IN)
    • Flutter version 2.13.0-0.0.pre.357 at /Users/mahesh/Documents/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7971b412c1 (3 days ago), 2022-04-01 05:31:08 -0400
    • Engine revision c3274494b0
    • Dart version 2.17.0 (build 2.17.0-261.0.dev)
    • DevTools version 2.12.1

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-32, build-tools 31.0.0
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! CocoaPods 1.10.2 out of date (1.11.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.65.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.37.20220301

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 12.3 21E230 darwin-arm
    • Chrome (web)    • chrome • web-javascript • Google Chrome 100.0.4896.60
    ! Device 192.168.1.2:5555 is offline.

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```
</details>

cc: @annagrin

**Comment by jason-simmons** [comment]
Reproduced this using the DevTools debugger in Chrome 100.0.4896.75

Looked at the events received by DWDS and noticed that the `url` field is empty in the `callFrames` within the `DebuggerPausedEvent` sent by Chrome 100.  Chrome 99 populates the `url` as expected.

`calculateDartFrameFor` therefore fails to identify the stack frame.  DWDS then sends a `PauseBreakpoint` event that is missing the data needed by the debugger UI.

@grouma

**Comment by jason-simmons** [comment]
Apparently Chromium deprecated the `CallFrame.url` field in the debugger protocol:

https://bugs.chromium.org/p/chromium/issues/detail?id=1270316

https://chromium.googlesource.com/v8/v8/+/21f72f82ce854762d6bc1c780d3b108dc1947f5d

**Comment by grouma** [comment]
Now that the paused event only has the `scriptId` we need to create a map from `scriptId` to `url` so that we can properly determine the corresponding Dart files. 

We explicitly [ignore](https://github.com/dart-lang/webdev/blob/bd3ab2fb5702319bb52f11323d1b1ab16867a8bb/dwds/debug_extension/web/background.dart#L578) the `scriptParsed` events, which are necessary for creating this mapping, in the extension for historical performance reasons.

If sending all of the `scriptParsed` events continues to be a performance issue we could create a synthetic API to look up the URL for a `scriptId`. In this scenario the `scriptParsed` events will be collected by the extension but not forwarded. 

The tricky part of this proposal is that users are not always debugging with the extension in the middle. However, sending `scriptParsed` events over a local debug connection did not have historical performance issues so a hybrid approach may be appropriate.

**Comment by annagrin** [comment]
This is the same as https://github.com/dart-lang/webdev/issues/1557

**Comment by annagrin** [comment]
Duplicate of:
https://github.com/flutter/flutter/issues/101318
https://github.com/flutter/flutter/issues/101139

**Comment by SaurabGhimire** [comment]
Thanks for filing this issue. I was stuck in this for 1 week.

**Comment by DanielArndt** [comment]
Now that [a fix has been merged to dart webdev tools](https://github.com/dart-lang/webdev/pull/1559), is there an easy way to test / back-port this? 

Upgrading flutter to use `dwds-v13.0.0` seems non-trivial to me, but I'm very new to flutter.

**Comment by annagrin** [comment]
> Now that https://github.com/dart-lang/webdev/pull/1559, is there an easy way to test / back-port this? 

> Upgrading flutter to use dwds-v13.0.0 seems non-trivial to me, but I'm very new to flutter.

@DanielArndt  I am working on upgrading flutter. This PR is bringing the new version of dwds to flutter master: https://github.com/flutter/flutter/pull/101450.

I am investigating if it is possible to hot-fix flutter stable as well, will update this thread when I know the answer.

If you want to test it locally, you would need a dev setup of flutter (see: https://github.com/flutter/flutter/blob/master/CONTRIBUTING.md), and then:

- `cd <path to flutter repo>/flutter/packages/flutter_tools`
- update dwds constraint in `packages/flutter_tools/pubspec.yaml` to `13.0.0`
- run `flutter pub get` in `packages/flutter_tools` directory
- remove your cache: `rm -rf ../../bin/cache`
- rebuild flutter: `flutter pub get`

Then you can proceed with your normal workflow (debug from VSCode, IntelliJ, or flutter run+DevTools).

**Comment by DanielArndt** [comment]
Thanks @annagrin !

The steps you gave above were similar to what I tried, except I used `stable` and it led to a cascade of upgrades in other packages. Using `master` went a lot smoother. Web debugging seems to be working.

Side note  -- mostly for others that may want to follow along (I know this is tangential to the problem being discussed): it wasn't quite clear to me in which directory the final `flutter pub get` should be run, since in the other steps you provided an explicit path. I ran it at both the top level (which has no `pubspec.yaml`) and in `packages/flutter_tools` which seemed to work, but I'm assuming the former was incorrect. -- Is the first `flutter pub get` really necessary if we're going to remove the cache? or did the `flutter pub get` in the top level directory recursively run it in, for example, `packages/flutter`?

**Comment by annagrin** [comment]
>  Is the first flutter pub get really necessary if we're going to remove the cache? or did the flutter pub get in the top level directory recursively run it in, for example, packages/flutter?

We are only using it to trigger the flutter build so it does not actually matter where you run it:)

**Comment by coastalgit** [comment]
In stable version flutter_tools is already using 11.5.2 and debugging is still not possible.

**Comment by annagrin** [comment]
> In stable version flutter_tools is already using 11.5.2 and debugging is still not possible.

Current flutter stable is using dwds version 11.5.1, I released a 11.5.2 hotfix yesterday and it is still in progress (See https://github.com/flutter/flutter/issues/101531). 

Are you modifying your flutter following the instructions I posted above?

**Comment by annagrin** [comment]
Not closed just yet!

Update: 
- **The issue is fixed in flutter master channel**!
- Pending:
  - hotfix to stable (https://github.com/flutter/flutter/issues/101531)
  - hotfix to next stable candidate (TBD)

**Comment by annagrin** [comment]
Thought I'd post a separate workaround for flutter stable to unblock users until the hotfix is submitted.
Note: no workaround is needed for flutter master, just a sync.

- cd `<path to flutter repo>/flutter/packages/flutter_tools`
- update dwds constraint in `packages/flutter_tools/pubspec.yaml` to `11.5.2`
- run `flutter pub get` in `packages/flutter_tools` directory
- remove your cache: `rm -rf ../../bin/cache`
- rebuild flutter: `flutter pub get  `packages/flutter_tools` directory (it triggers the flutter rebuild)

Then you can proceed with your normal workflow (debug from VSCode, IntelliJ, or flutter run+DevTools).

**Comment by fhuonder** [comment]
Hi @annagrin 
At least for me the workaround is not working... When the debugger hits the breakpoint it stops but without showing my the highlighted line nor variables or any other context.
In the console it says:
```
Debugger: Failed to create dart frame for [_myMethod]: cannot find location for script 237
Debugger: Error calculating Dart framesError: NoSuchMethodError: The getter 'length' was called on null.
Receiver: null
Tried calling: length
Debugger: Failed to create dart frame for [_myParentMethod]: cannot find location for script 237
Debugger: Failed to create dart frame for : cannot find location for script 237
Debugger: Failed to create dart frame for build: cannot find location for script 146
Debugger: Failed to create dart frame for build: cannot find location for script 35
```
The method `_myMethod` contains the breakpoint and `_myParentMethod` calls `_myMethod`.

**Comment by coastalgit** [comment]
> > In stable version flutter_tools is already using 11.5.2 and debugging is still not possible.
> 
> Current flutter stable is using dwds version 11.5.1, I released a 11.5.2 hotfix yesterday and upgrading flutter stable to it is still in progress (See #101531).
> 
> Are you modifying your flutter following the instructions I posted above?

My apologies - I had not flushed the caches! I can confirm I applied dwds 11.5.2 in flutter_tools and my flutter web project is hitting its breakpoints (in Android Studio) correctly in chrome v100.0.4896.7.

**Comment by annagrin** [comment]
@fhuonder can you please share a repro?

**Comment by Efremov-Evgeny** [comment]
After create new project, have new error

Launching lib[/main.dart]() on Chrome in debug mode...
lib/main.dart:1
**DartUri: Unresolved uri: dart:web_sql
DartUri: Unresolved uri: dart:ui**
This app is linked to the debug service: ws://127.0.0.1:55141[/QCQRRvvvoxE]()=[/ws]()
Debug service listening on ws://127.0.0.1:55141[/QCQRRvvvoxE]()=[/ws]()
💪 Running with sound null safety 💪
Connecting to VM Service at ws://127.0.0.1:55141[/QCQRRvvvoxE]()=[/ws]()
Flutter Web Bootstrap: Programmatic

**Comment by renanyoy** [comment]
when the fix should be available on stable channel ?

**Comment by fhuonder** [comment]
> @fhuonder can you please share a repro?

Sorry, unfortunately not. This is an internal project.

**Comment by annagrin** [comment]
@fhuonder does this issue repro for you on a counter project?

**Comment by fhuonder** [comment]
Hi @annagrin 

All breakpoints do not work.

I created a new project (`flutter create`). On Flutter 2.10.4 with the workaround (https://github.com/flutter/flutter/issues/101224#issuecomment-1090830539) applied and I encounter the same error.
What I see:

- Breakpoint in `main.dart` on the `_counter++` line: The debugger stops but with no visual stopping in vscode.
- Breakpoint in `main.dart` on the `return Scaffold(` line in the `build` method: The debugger stops in `binding.dart` on `window.scheduleFrame();` (line 790)

The I switched to master (verison `2.13.0.-0.0.pre.523`).
Unfortunately, there I cannot start the application because I hit this error here https://github.com/flutter/flutter/issues/101639 although it should be fixed on master.  But the interesting thing here is when I create a breakpoint somewhere in the framework code (before the app crashes) the breakpoint stops. Would be better if it stopped in the non-framework code.

My environment (flutter doctor from master channel):
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, 2.13.0-0.0.pre.523, on Ubuntu 20.04.4 LTS 5.4.72-microsoft-standard-WSL2, locale en_US.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0-rc3)
[✗] Chrome - develop for the web (Cannot find Chrome executable at google-chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.
[!] Android Studio (not installed)
[✓] Connected device (1 available)
[✓] HTTP Host Availability

! Doctor found issues in 2 categories.
```

I work with vscode and a devcontainer (Windows 10 with WSL2 and docker is WSL2-based).
Therefore, I need to use the `web server device`, maybe this has an influence?

Regards,
Florian

**Comment by annagrin** [comment]
@fhuonder Issue https://github.com/flutter/flutter/issues/101639 is just printing, should not affect the way the app runs. I'll make sure to remove the confusing printing in the future. I suspect your issue is due to using web server device, so I suggest to create a separate issue for that so we can address it separately.

**Comment by fhuonder** [comment]
@annagrin
Can you reproduce it with the web server device?

**Comment by annagrin** [comment]
I am not familiar with all the flutter devices - could you please add some instructions on your repro?

**Comment by annagrin** [comment]
@fhuonder I can repro this issue now, opened https://github.com/dart-lang/webdev/issues/1581 to track the fix.

**Comment by JaredEzz** [comment]
This issue was addresses in the hotfix v2.10.5 of the stable branch 

https://github.com/flutter/flutter/wiki/Hotfixes-to-the-Stable-Channel#2105-april-18-2022

https://twitter.com/FlutterReleases/status/1516140054183387140?t=ljVpMAzd2-VfYhYeB1S52w&s=19

**Comment by maheshmnj** [comment]
I can confirm the fix has landed in `stable 2.10.5`

<img width="1418" alt="image" src="https://user-images.githubusercontent.com/31410839/163923945-10402ff6-95ab-44a1-a0de-d47194a94b5c.png">

```
Flutter 2.10.5 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 5464c5bac7 (12 hours ago) • 2022-04-18 09:55:37 -0700
Engine • revision 57d3bac3dd
Tools • Dart 2.16.2 • DevTools 2.9.2
```

**Comment by fzyzcjy** [comment]
This issue is in https://github.com/flutter/flutter/wiki/Hotfixes-to-the-Stable-Channel so I guess it can be closed 🎉

**Comment by tgpsantos** [comment]
I've tested in the 2.10.5 and it's working

**Comment by maheshmnj** [comment]
I am going to close this issue based on the above reports. In case anyone encounters this issue again feel free to file a new issue or comment on this issue and we will reopen it.

Thank you.

**Comment by alpiepho** [comment]
Thank for fixing this!

**Comment by fabidona-mikamai** [comment]
thank you for the fix guys!

**Comment by annagrin** [comment]
@fhuonder 

The web server issue fix still in progress, I created the following issue to track its progress separately: https://github.com/flutter/flutter/issues/102159

**Comment by ugurberkecan** [comment]
any update ?

**Comment by JaredEzz** [comment]
> any update ?

Yes, this is fixed. Make sure you're on the latest version of Chrome and the stable branch of Flutter

**Comment by annagrin** [comment]
@fabianogadenz could you  please open a separate issue for the problem you are encountering, with repro steps? I suspect it is caused by a different issue from the one that was fixed in the current one. That would let us investigate and track the issue separately.
